### PR TITLE
Improve Microsoft Visual Studio CI & fix build on MSVC 2015,2017

### PR DESCRIPTION
**Pull request by phprus**
New CI:
- Toolset Visual Studio 2015 (Win32, x64)
- Toolset Visual Studio 2017 (Win32, x64)
- Toolset Visual Studio 2019 (Win32, x64)
- Toolset Visual Studio 2022 (Win32, x64)

Fix build errors on MSVC 2015,2017.

cc @mtl1979

### CODE REVIEW

File: .github/workflows/cmake.yml [hunk 1]
```diff
@@ -348,15 +348,45 @@ jobs:
...
 
-          - name: Windows MSVC Win32
+          - name: Windows MSVS 2022 v143 Win32
             os: windows-latest
             compiler: cl
-            cmake-args: -A Win32
+            cmake-args: -G "Visual Studio 17 2022" -A Win32 -T v143
 
-          - name: Windows MSVC Win64
+          - name: Windows MSVS 2022 v143 Win64
```
Comment (mtl1979) [inline .github/workflows/cmake.yml]: Should use `MSVC` as it is more accurate... ´VS` means Visual Studio and it is frontend for multiple programming languages. VC can mean either C, C++ or C#, but these three languages are handled by same set of compiler binaries, commonly known as Common Language (CL) compiler.

Comment (phprus) [reply .github/workflows/cmake.yml]: Renamed.

File: fallback_builtins.h [hunk 2]
```diff
@@ -71,6 +71,24 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
...
+/* Missing zero-extension AVX and AVX512 intrinsics.
+ * Fixed in version 15.7
+ * https://developercommunity.visualstudio.com/t/missing-zero-extension-avx-and-avx512-intrinsics/175737
+ */
+#if defined(_MSC_VER) && _MSC_VER < 1914
+#ifdef X86_AVX2
+static inline __m256i _mm256_zextsi128_si256(__m128i a) {
+    return _mm256_inserti128_si256(_mm256_setzero_si256(), a, 0);
+}
+#endif // X86_AVX2
```
Comment (mtl1979) [inline fallback_builtins.h]: This gives a warning on line 81 complaining that `/arch:AVX` isn't used. Maybe test for `__AVX2__` to see the code doesn't get included without proper compiler option.

Comment (phprus) [reply fallback_builtins.h]: zlib-ng buildsystem does not set ``/arch:AVX*`` on MSVC.
I'll fix it in a separate commit.

File: fallback_builtins.h [hunk 3]
```diff
@@ -71,6 +71,24 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
...
+
+#ifdef X86_AVX512
+static inline __m512i _mm512_zextsi128_si512(__m128i a) {
+    return _mm512_inserti32x4(_mm512_setzero_si512(), a, 0);
+}
+#endif // X86_AVX512
```
Comment (mtl1979) [inline fallback_builtins.h]: Should most likely test for `__AVX512F__` like with gcc/clang to avoid warning.

Comment (phprus) [reply fallback_builtins.h]: Fixed.

File: fallback_builtins.h (hunk 2 above)
Comment (phprus) [reply fallback_builtins.h]: Fixed.

**Comment by mtl1979** [comment]
Looks a lot better now with v140 on both x86 and x64...

**Comment by nmoinvaz** [comment]
This is a great idea! Is it possible to get VS2013 support through Chocolately?

**Comment by phprus** [comment]
@nmoinvaz
Theoretically yes. But I'm not sure.
I think VS2013 support should be developed in a separate PR after this PR is merged.

**Comment by mtl1979** [comment]
I'm inclined to say the implicit padding warnings, and sign mismatch warnings in some tests, can be fixed in another PR if them annoy some people too much...

**Review by nmoinvaz — APPROVED** [review]

**Review by mtl1979 — APPROVED** [review]

File: fallback_builtins.h [hunk 4]
```diff
@@ -71,6 +71,24 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
 
 #endif // __AVX2__
 
+/* Missing zero-extension AVX and AVX512 intrinsics.
+ * Fixed in version 15.7
```
Comment (Dead2) [inline fallback_builtins.h]: This should specify version 15.7 of what exactly, not that it is impossible to understand without it specified, it just makes it easier to understand when doing a quick read of the comment.

Comment (phprus) [reply fallback_builtins.h]: @Dead2 
Comment updated.

Comment (Dead2) [reply fallback_builtins.h]: Excellent 👍

**Review by Dead2 — APPROVED** [review]
LGTM

**Comment by Dead2** [comment]
This needs a rebase after other PRs were merged, would you please rebase @phprus?

**Comment by phprus** [comment]
@Dead2
Rebased

**Pull request merged**